### ISSUE TITLE
Adress sanitizer detects a lot of leaks during some tests

### ISSUE THREAD

**Issue by Project579** [David-Haim/concurrencpp#130]
When running the tests with AddressSanitizer a large amount of leaked memory is detected, here are the logs for the relevant failed tests:
[asan_failures.log](https://github.com/David-Haim/concurrencpp/files/11482374/asan_failures.log)

System Information:
OS: Arch Linux x86_64
Compiler: clang version 15.0.7
Compilation flags:
```
CXXFLAGS="-fsanitize=address,leak,undefined"
LDFLAGS="-fsanitize=address,leak,undefined"
```

**Update:**
My guess is that these leaks are caused by a recursion with some `shared_pointers` or similar.

**Comment by David-Haim** [comment]
So, only 'shared_result' is leaking?
Can you re-run the asan tests with 'develop' branch?

**Comment by Project579** [comment]
I have re-run the tests using the develop branch, here is the resulting log:
[asan_failures.log](https://github.com/David-Haim/concurrencpp/files/11491900/asan_failures.log)

and yes, only the "shared_result" tests are failing.

**Comment by David-Haim** [comment]
I have this feeling that this is somewhat caused by clang-15 compilation bug.
Can you please run this again with an older clang version?

**Comment by David-Haim** [comment]
fixed with #139
